src/Renode/Network/NetworkServer/Modules/IcmpServerModule.cs

```diff
@@ -0,0 +1,177 @@
+            if (ipv4Packet.DestinationAddress.Equals(IP))
+            {
+                this.Log(LogLevel.Warning, "Wrong destination address: {0}",
+                    ipv4Packet.DestinationAddress);
+                return false;
+            }
+
+            // For now we only respond to Echo Requests so everything else is discarded
+            if (!((ICMPv4Packet)ipv4Packet.PayloadPacket).TypeCode.Equals(ICMPv4TypeCodes.EchoRequest))
+            {
+                this.Log(LogLevel.Warning, "Unsupported ICMP code: {0}",
+                    ((ICMPv4Packet)ipv4Packet.PayloadPacket));
+                return false;
+            }
+
+            ICMPv4TypeCodes.EchoReply.AsRawBytes().CopyTo(byteReply, 0);
+            return true;
+        }
+
+
+        /// <summary>
+        /// Creates an ICMPv4 packed with a given response
+        /// </summary>
+        /// <param name="ipv4Packet"></param>
+        /// <param name="byteReply"></param>
+        /// <returns></returns>
+        private ICMPv4Packet CreateIcmpv4Packet(IPv4Packet ipv4Packet, byte[] byteReply)
+        {
+            var icmpv4Packet = (ICMPv4Packet)ipv4Packet.PayloadPacket;
+
+            var icmpv4PacketResponse = new ICMPv4Packet(new ByteArraySegment(byteReply));
+
+            // We can copy that from request packet because they are the same in the request and the replay
+            icmpv4Packet.Data.CopyTo(icmpv4PacketResponse.Data, 0);
+            icmpv4PacketResponse.ID = icmpv4Packet.ID;
+            icmpv4PacketResponse.Sequence = icmpv4Packet.Sequence;
+
+            return icmpv4PacketResponse;
+        }
+
+
+        /// <summary>
+        /// Creates an IPv4 response packet
+        /// </summary>
+        /// <param name="ipv4Packet"></param>
+        /// <returns></returns>
+        private IPv4Packet CreateIPv4Packet(IPv4Packet ipv4Packet)
+        {
+            var ipv4PacketResponse = new IPv4Packet(IP,
+                ((IPv4Packet)ipv4Packet.ParentPacket).SourceAddress);
+            return ipv4PacketResponse;
+        }
+
+
+        /// <summary>
+        /// Creates an Ethernet Frame from a given IPv4, and ICMPv4 packet
+        /// </summary>
+        /// <param name="ipv4Packet"></param>
+        /// <param name="icmpv4PacketResponse"></param>
+        /// <param name="icmpDestinationAddress"></param>
+        /// <returns></returns>
+        private EthernetFrame CreateEthernetFramePacket(IPv4Packet ipv4Packet, ICMPv4Packet icmpv4PacketResponse,
+            PhysicalAddress icmpDestinationAddress)
+        {
+            ipv4Packet.PayloadPacket = icmpv4PacketResponse;
+
+            var ethernetResponse = new EthernetPacket((PhysicalAddress)MAC,
+                icmpDestinationAddress,
+                EthernetPacketType.None)
+            {
+                PayloadPacket = ipv4Packet
+            };
+            icmpv4PacketResponse.UpdateCalculatedValues();
+
+            this.Log(LogLevel.Noisy, "Sending response: {0}",
```

Comment:
in general, logging via `this` is not recommended here. Your sever has no name in Renode, so the logs will have no owner. It's better to pass the NetworkServer as a constructor parameter to IcmpServerModule, store it in a field (typically called `parent`) and log with `parent.Log`